### PR TITLE
Add iCE40 boards flashing support

### PR DESCRIPTION
**Pull request by smunaut**
(also contains misc fixes / tweak to actually make it work)

### CODE REVIEW

**Comment by whitequark** [comment]
There already is a `program-ice40` applet. Have you looked at it?

**Comment by smunaut** [comment]
AFAICT the `program-ice40` is for SRAM loading directly right ?

This is to program the flash of ice40 boards where you basically need to assert the reset line of the fpga (so you can drive the SPI lines) and then program it like any other SPI flash, then release reset to let the fpga program itself.

File: software/glasgow/applet/spi_master/__init__.py [hunk 1]
```diff
@@ -54,10 +54,11 @@ def __init__(self, pads, sck_idle, sck_edge, ss_active):
...
 
-CMD_XFER    = 0x00
-CMD_READ    = 0x01
-CMD_WRITE   = 0x02
-BIT_HOLD_SS = 0x80
+CMD_XFER     = 0x00
+CMD_READ     = 0x01
+CMD_WRITE    = 0x02
+CMD_GPIO     = 0x08
+BIT_HOLD_SS  = 0x80
```
Comment (whitequark) [inline software/glasgow/applet/spi_master/__init__.py]: Adding a GPIO command is not the right way to do this--the right way would be to add an I2C register in the leaf applet (in this case, `spi-flash-ice40`). That is much cleaner and more flexible, too.

File: software/glasgow/applet/spi_flash_25c/__init__.py [hunk 2]
```diff
@@ -350,7 +352,7 @@ def _show_progress(done, total, status):
                 await flash_iface.read_manufacturer_device_id()
             long_manufacturer_id, long_device_id = \
                 await flash_iface.read_manufacturer_long_device_id()
-            if long_manufacturer_id == manufacturer_id:
+            if long_manufacturer_id == manufacturer_id or manufacturer_id == 0xff:
```
[the remark this answers is not in the document]
Comment (smunaut) [reply software/glasgow/applet/spi_flash_25c/__init__.py]: On the couple of flash I tried none supported the 0x90 command at all and just answered 0xff.

In the current code that meant that it was displaying "0xff (unknown)" as the identify response even though the long identifier was valid.

File: software/glasgow/applet/spi_master/__init__.py (hunk 1 above)
Comment (smunaut) [reply software/glasgow/applet/spi_master/__init__.py]: Is there an example of an applet doing this ?

File: software/glasgow/applet/spi_flash_25c/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (whitequark) [reply software/glasgow/applet/spi_flash_25c/__init__.py]: Thanks. Please add a comment to this effect.

File: software/glasgow/applet/spi_master/__init__.py (hunk 1 above)
Comment (whitequark) [reply software/glasgow/applet/spi_master/__init__.py]: Yes, `spi-flash-avr`.

Comment (smunaut) [reply software/glasgow/applet/spi_master/__init__.py]: Yes thanks indeed it looks both much cleaner and much easier, I had no idea this infrastructure existed !

File: software/glasgow/applet/spi_flash_ice40/__init__.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        return SPIFlashICE40Interface(flash_iface, self.logger, device,
+                                      self.__addr_dut_reset, self.__addr_dut_done)
+
+    async def interact(self, device, args, ice40_iface):
+        await ice40_iface.set_reset(True)
+        await super().interact(device, args, ice40_iface.lower)
+        await ice40_iface.set_reset(False)
+
+        if args.pin_done:
+            time.sleep(0.1)
```
[the remark this answers is not in the document]
Comment (whitequark) [reply software/glasgow/applet/spi_flash_ice40/__init__.py]: Ah no, that's fine. Just needs a comment.

**Comment by smunaut** [comment]
Any chance you can have a quick look at the updated patch ?

**Comment by whitequark** [comment]
Sorry! I've looked at it but forgot to write it down. Mainly, I have not realized before that this is an applet for flashing iCE40 *boards* with *regular SPI flash on them*. (I thought it was for flashing NVM on iCE40 FPGAs themselves.) I got confused because of the way you named the applet.

Two things:
  1. As per the paragraph above, `spi-flash-ice40` is definitely a bad name for this applet. But I am not sure what would be a good name. Ideas welcome! I think @gregdavill has some similar contraption for ECP5 which would have a similar name conundrum.
  2. I think you should poll CDONE pin, in software, with a timeout (let's say, generously, 2 seconds). This feels like a more correct approach than just a delay and a check. Do you agree?

**Comment by gregdavill** [comment]
I'm writing the ecp5 equivalent of ```progarm-ice40```. So for now I've just named mine ```program-ecp5```. It sends all commands over JTAG, with the intent to handle SRAM bitstream loading too. But the first feature I'm working on is the JTAG ⇌ SPI passthorugh feature of the ECP5 to program SPI FLASH.

Wolud it be appliciable to bundle this into ```program-ice40``` as new operations? Or do you think the functionality is too different for that to make sense?

Otherwise naming wise are ```program-ice40-25c```,```program-ice40-spi-flash```, ```program-ice40-flash``` any better?

**Comment by mithro** [comment]
Random dry-by thoughts....

How about something like `program-spi-via-ice40proxy` and `program-spi-via-ecp5proxy`? Other options could be`through` (which is kind of long) - maybe just something like `by` or `using` could also work? Dunno if you need the proxy part?

I imagine for doing something like programming the SPI flash on a Xilinx board through the JTAG interface using a gateware proxy could be named something like - `program-spi-via-jtag-via-xc7proxy` could work?

**Comment by whitequark** [comment]
I think it would make sense to have the following naming:
  * What is currently called `program-ice40` would be `program-ice40-sram`
  * This applet would be called `program-ice40-flash`
  * A hypothetical NVM applet would be called `program-ice40-nvm`
  * @gregdavill's applet would be called `program-ecp5-jtag-flash`.

This isn't perfect nor it unambiguously explains what's going on. That's fine. We have the applet descriptions for that. It just needs to be easy to remember and map well to actual operation.

**Comment by gregdavill** [comment]
This naming makes sense to me. 

It's verbose enough to differentiate functionality.

File: software/glasgow/applet/program_ice40_sram/__init__.py [hunk 4]
```diff
@@ -117,11 +117,12 @@ def __init__(self, pads, out_fifo, sys_clk_freq):
...
 
-class ProgramICE40Applet(GlasgowApplet, name="program-ice40"):
+class ProgramICE40SRAMApplet(GlasgowApplet, name="program-ice40-sram"):
     logger = logging.getLogger(__name__)
-    help = "program iCE40 FPGAs"
+    help = "program iCE40 FPGAs SRAM"
     description = """
-    Program iCE40 FPGAs.
+    Program iCE40 FPGAs SRAM. This directly loads a bitstream into the FPGA
+	volatile memory.
```
Comment (whitequark) [inline software/glasgow/applet/program_ice40_sram/__init__.py]: Broken indentation.

File: software/glasgow/applet/program_ice40_flash/__init__.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    def add_build_arguments(cls, parser, access):
+        super().add_build_arguments(parser, access)
+
+        access.add_pin_argument(parser, "reset")
+        access.add_pin_argument(parser, "done")
+
+    def build(self, target, args):
+        subtarget = super().build(target, args)
+
+        if args.pin_reset:
```
[the remark this answers is not in the document]
Comment (smunaut) [reply software/glasgow/applet/program_ice40_flash/__init__.py]: Fixed. Also added a commit fixing the same but from spi_flash applet where I copied this from.

File: software/glasgow/applet/program_ice40_flash/__init__.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    def __init__(self, interface, logger, device, addr_dut_reset, addr_dut_done):
+        self.lower   = interface
+        self._logger = logger
+        self._level  = logging.DEBUG if self._logger.name == __name__ else logging.TRACE
+        self._device = device
+        self._addr_dut_reset = addr_dut_reset
+        self._addr_dut_done  = addr_dut_done
+
+    async def set_reset(self, reset):
+        if self._addr_dut_reset:
```
[the remark this answers is not in the document]
Comment (smunaut) [reply software/glasgow/applet/program_ice40_flash/__init__.py]: fixed

File: software/glasgow/applet/program_ice40_flash/__init__.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+            await self._device.write_register(self._addr_dut_reset, int(reset))
+
+    async def get_done(self):
+        if self._addr_dut_reset:
+            return await self._device.read_register(self._addr_dut_done)
+
+
+class ProgramICE40FlashApplet(SPIFlash25CApplet, name='program-ice40-flash'):
+    logger = logging.getLogger(__name__)
+    help = "Programs the 25Cxx flash chip on ice40 boards"
```
[the remark this answers is not in the document]
Comment (smunaut) [reply software/glasgow/applet/program_ice40_flash/__init__.py]: Right. Rewrote it to be more consistent and try to use same terminology.

**Comment by whitequark** [comment]
Thanks for the contribution! I've merged it manually with some further adjustments to prose for clarity.

**Pull request closed without merging**